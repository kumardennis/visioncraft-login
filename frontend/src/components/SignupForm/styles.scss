h2 {
  color: white;
}
.signup-form {
  position: relative;
  background: linear-gradient(
    to bottom right,
    rgba(0, 0, 0, 0.8),
    rgba(25, 42, 56, 0.8)
  );
  padding: 2rem;
  width: 400px;
  height: fit-content;
  box-shadow: 0 0 20px 10px black;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;

  margin: auto;

  @media (max-width: 1024px) {
    width: 90%;
    margin: 2rem 0;
  }

  form {
    display: flex;
    flex-direction: column;

    label {
      margin: 3rem 0 1rem 0;
      z-index: 10;
      color: white;
      font-size: larger;
      border: none;
      width: fit-content;
    }

    .user-input {
      padding: 1.5rem;
      border: 2px solid lightgrey;
      border-radius: 1rem;
      background-color: transparent;
      color: white;

      &::placeholder {
        color: rgb(107, 107, 107);
      }

      &:focus {
        outline: none;
      }
    }

    #submit-sign-up {
      background: linear-gradient(to bottom right, #00a2ff, #0e6597);
      box-shadow: 0 0 5px darkblue;
      border: none;
      outline: none;
      padding: 1rem;
      margin: 2rem 0;
      color: white;
      font-weight: bold;
      border-radius: 50px;
    }
  }

  .toc {
    margin: 2rem;
    color: lightgrey;
    text-align: center;
  }
}
