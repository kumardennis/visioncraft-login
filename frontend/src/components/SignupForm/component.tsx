import React, { useState, useEffect } from "react";
import { MdEmail } from "react-icons/md";
import { AiTwotoneLock } from "react-icons/ai";
import { BsArrowBarRight } from "react-icons/bs";
import "./styles.scss";

interface SignupFormProps {
  changeAuthForm: () => void;
}

interface SignupData {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}

interface Response {
  type: string;
  message: string;
}

const SignupForm = ({ changeAuthForm }: SignupFormProps) => {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [signupData, setSignupData] = useState<SignupData | {}>({});

  const handleInput = (event: React.FormEvent<HTMLInputElement>) => {
    switch (event.currentTarget.id) {
      case "user-first-name":
        setFirstName(event.currentTarget.value);
        break;

      case "user-last-name":
        setLastName(event.currentTarget.value);
        break;

      case "user-email":
        setEmail(event.currentTarget.value);
        break;

      case "user-password":
        setPassword(event.currentTarget.value);
        break;
      default:
        break;
    }
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    setSignupData({
      firstName,
      lastName,
      email,
      password,
    });
  };

  useEffect(() => {
    if (Object.keys(signupData).length !== 0) {
      fetch("/signup", {
        method: "PUT",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(signupData),
      })
        .then((response) => response.json())
        .then((data: Response) => {
          alert(data.message); // TODO: add Toastify
          data.type === "success" && changeAuthForm();
        })
        .catch((err: Error) => alert(err)); // TODO: add Toastify
    }
  }, [signupData, changeAuthForm]);

  return (
    <div className="signup-form">
      <h2>Sign up!</h2>
      <form onSubmit={handleSubmit}>
        <label htmlFor="user-first-name">First Name</label>
        <input
          className="user-input"
          id="user-first-name"
          name="user-first-name"
          type="text"
          placeholder="Eg: John"
          onChange={handleInput}
          value={firstName}
          required
        />

        <label htmlFor="user-last-name">Last Name</label>
        <input
          className="user-input"
          id="user-last-name"
          name="user-last-name"
          type="text"
          placeholder="Eg: Holden"
          onChange={handleInput}
          value={lastName}
          required
        />

        <label htmlFor="user-email">
          Email <MdEmail />
        </label>
        <input
          className="user-input"
          id="user-email"
          name="user-email"
          type="email"
          placeholder="Eg: abc123@gmail.com"
          onChange={handleInput}
          value={email}
          required
        />

        <label htmlFor="user-password">
          Password <AiTwotoneLock />
        </label>
        <input
          className="user-input"
          id="user-password"
          name="user-password"
          type="password"
          placeholder="Eg: somecrazyAs123Password"
          onChange={handleInput}
          value={password}
          required
        />

        <input id="submit-sign-up" type="submit" value="Sign up!" />
      </form>

      <div className="toc">
        Use of this service is subject to terms and conditions.
      </div>

      <button onClick={changeAuthForm} className="change-auth-screen">
        Take me to Sign In! <BsArrowBarRight />
      </button>
    </div>
  );
};

export default SignupForm;
