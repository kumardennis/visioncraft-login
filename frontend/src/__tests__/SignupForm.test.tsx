import React from "react";
import { mount, shallow, configure } from "enzyme";
import SignupForm from "../components/SignupForm/component";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("Signup Form", () => {
  it("should render SignupForm", () => {
    const signupForm = shallow(<SignupForm handleIsSigningUp={() => {}} />);
    expect(signupForm).toMatchSnapshot();
  });
});

describe("Inputs for SignupForm", () => {
  let wrapper;

  test("Render firstName check", () => {
    wrapper = mount(<SignupForm handleIsSigningUp={() => {}} />);
    const handleInput = jest.fn();

    const firstNametInput = wrapper
      .find('input[id="user-first-name"]')
      .simulate("change");

    expect(firstNametInput).toMatchSnapshot();
  });

  test("Render lastName check", () => {
    wrapper = mount(<SignupForm handleIsSigningUp={() => {}} />);
    const handleInput = jest.fn();

    const lastNametInput = wrapper
      .find('input[id="user-last-name"]')
      .simulate("change");

    expect(lastNametInput).toMatchSnapshot();
  });

  test("Render email check", () => {
    wrapper = mount(<SignupForm handleIsSigningUp={() => {}} />);
    const handleInput = jest.fn();

    const emailInput = wrapper
      .find('input[id="user-email"]')
      .simulate("change");

    expect(emailInput).toMatchSnapshot();
  });

  test("Render password check", () => {
    wrapper = mount(<SignupForm handleIsSigningUp={() => {}} />);
    const handleInput = jest.fn();

    const passwordInput = wrapper
      .find('input[id="user-password"]')
      .simulate("change");

    expect(passwordInput).toMatchSnapshot();
  });
});
