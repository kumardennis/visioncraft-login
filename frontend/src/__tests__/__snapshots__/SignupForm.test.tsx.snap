// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inputs for SignupForm Render email check 1`] = `ReactWrapper {}`;

exports[`Inputs for SignupForm Render firstName check 1`] = `ReactWrapper {}`;

exports[`Inputs for SignupForm Render lastName check 1`] = `ReactWrapper {}`;

exports[`Inputs for SignupForm Render password check 1`] = `ReactWrapper {}`;

exports[`Signup Form should render SignupForm 1`] = `ShallowWrapper {}`;
